### PR TITLE
Vim emulation: add "o", "O" commands

### PR DESCRIPTION
**Pull request by mcclurgm**
Implements a basic form of the `o` and `O` commands in vim. It adds a new line before or after the current, navigates to that line, and enters insert mode.

As with #712 (and the other insertion commands), this doesn't properly work with numbers. It won't repeat the insert multiple times, only once.

### CODE REVIEW

File: plugins/vim-emulation/vim-emulation.vala [hunk 1]
```diff
@@ -205,6 +205,33 @@ public class Scratch.Plugins.VimEmulation : Peas.ExtensionBase, Peas.Activatable
...
+                    return false;
+                }
+                mode = Mode.INSERT;
+                debug ("Vim Emulation: INSERT Mode!");
+
+                view.move_cursor (Gtk.MovementStep.PARAGRAPH_ENDS, 1, false);
+                var buffer = view.buffer;
+                Gtk.TextIter insert;
+                buffer.get_iter_at_mark (out insert, buffer.get_insert ());
+                buffer.insert (ref insert, "\n", 1);
```
Comment (jeremypw) [inline plugins/vim-emulation/vim-emulation.vala:219]: Can't this be replaced by view.insert_at_cursor ("\n"); ?

File: plugins/vim-emulation/vim-emulation.vala [hunk 2]
```diff
@@ -205,6 +205,33 @@ public class Scratch.Plugins.VimEmulation : Peas.ExtensionBase, Peas.Activatable
...
+                    return false;
+                }
+                mode = Mode.INSERT;
+                debug ("Vim Emulation: INSERT Mode!");
+
+                view.move_cursor (Gtk.MovementStep.PARAGRAPH_ENDS, -1, false);
+                var buffer = view.buffer;
+                Gtk.TextIter insert;
+                buffer.get_iter_at_mark (out insert, buffer.get_insert ());
+                buffer.insert (ref insert, "\n", 1);
```
Comment (jeremypw) [inline plugins/vim-emulation/vim-emulation.vala:232]: Also this.

**Review by jeremypw — CHANGES_REQUESTED** [review]
This works as expected but some simplification is possible.

File: plugins/vim-emulation/vim-emulation.vala (hunk 1 above)
Comment (mcclurgm) [reply plugins/vim-emulation/vim-emulation.vala]: Yep, seems to work for me. I don't think I knew about that before. Thanks!

**Comment by mcclurgm** [comment]
Ok, I fixed your requests. There's one big issue remaining here that I just discovered, that the `MovementStep.PARAGRAPH_START` goes to the beginning of the line excluding whitespace. So the shift-O command needs some more work. I'll see if I can fix that.

**Comment by mcclurgm** [comment]
Since the effect is pretty much the same as inserting a newline after the `0` motion, I'm going to wait until we figure that out and then follow that solution here.

File: plugins/vim-emulation/vim-emulation.vala [hunk 3]
```diff
@@ -205,6 +205,27 @@ public class Scratch.Plugins.VimEmulation : Peas.ExtensionBase, Peas.Activatable
...
+                view.insert_at_cursor ("\n");
+                break;
+            case Gdk.Key.O:
+                if (mode == Mode.INSERT) {
+                    return false;
+                }
+                mode = Mode.INSERT;
+                debug ("Vim Emulation: INSERT Mode!");
+
+                view.move_cursor (Gtk.MovementStep.PARAGRAPH_ENDS, -1, false);
```
Comment (jeremypw) [inline plugins/vim-emulation/vim-emulation.vala:225]: I think you can insert `view.move_cursor (Gtk.MovementStep.DISPLAY_LINE_ENDS, -1, false);` here to ignore whitespace.

**Comment by mcclurgm** [comment]
One more thought I just had. This doesn't respect the current line ending setting. I guess I'll have to fix that too…

**Comment by jeremypw** [comment]
Need someone to fix the flatpak CI before this can be merged.

**Review by jeremypw — APPROVED** [review]

**Comment by jeremypw** [comment]
I think any issues with non-standard line-endings can be dealt with, if required, later. Code is focused on code development on Linux.

**Pull request merged**